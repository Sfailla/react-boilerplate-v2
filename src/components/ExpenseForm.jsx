import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export class ExpenseForm extends Component {
	constructor(props) {
		super(props);

		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calenderFocused: false,
			error: ''
		};
	}

	onDescriptionChange = evt => {
		const description = evt.target.value;
		this.setState(() => ({ description }));
	};

	onNoteChange = evt => {
		const note = evt.target.value;
		this.setState(() => ({ note }));
	};

	onAmountChange = evt => {
		const amount = evt.target.value;
		// this regex only allows numbers in the input field.
		// also only lets you put 2 values after decimal
		// (basically it formats the number value to money syntax)
		if (amount.match(/^\d*(\.\d{0,2})?$/g)) {
			this.setState(() => ({ amount }));
		}
	};

	onDateChange = createdAt => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};

	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calenderFocused: focused }));
	};

	onSubmit = evt => {
		evt.preventDefault();

		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({
				error: 'Please provide description and amount'
			}));
		} else {
			this.setState(() => ({ error: '' }));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	};

	render() {
		return (
			<div>
				{this.state.error && (
					<p style={{ color: 'red' }}>-{this.state.error}-</p>
				)}
				<form onSubmit={this.onSubmit}>
					<input
						type='text'
						placeholder='Description'
						autoFocus
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input
						name='amount'
						type='text'
						onChange={this.onAmountChange}
						value={this.state.amount}
						placeholder='Amount'
					/>

					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calenderFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						onChange={this.onNoteChange}
						value={this.state.note}
						placeholder='Add a note for your expense (optional)'
					/>

					<button type='submit'>Add Expense</button>
				</form>
			</div>
		);
	}
}

export default ExpenseForm;
