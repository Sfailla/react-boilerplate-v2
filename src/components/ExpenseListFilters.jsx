import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters';

const ExpenseListFilters = (props) => (
	<div>
		<input
			style={{ height: '25px' }}
			type="text"
			value={props.filters.text}
			onChange={(evt) => {
				props.dispatch(setTextFilter(evt.target.value));
			}}
		/>
		<select
			value={props.filters.sortBy}
			onChange={(evt) => {
				evt.target.value === 'date' ? props.dispatch(sortByDate()) : props.dispatch(sortByAmount());
			}}
		>
			<option value="date">Date</option>
			<option value="amount">Amount</option>
		</select>
	</div>
);

const mapStateToProps = (state) => ({
	filters: state.filters
});

console.log();

export default connect(mapStateToProps)(ExpenseListFilters);
