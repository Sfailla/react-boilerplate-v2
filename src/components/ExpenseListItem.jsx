import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ dispatch, description, amount, createdAt, id }) => (
	<div>
		<h3>
			<Link to={`/edit/${id}`}>{description}</Link>
		</h3>
		<p>Amount: {amount}</p>
		<p>Created At: {createdAt}</p>
		<button
			onClick={() => {
				dispatch(removeExpense({ id }));
			}}
		>
			Remove
		</button>
	</div>
);

export default connect()(ExpenseListItem);
