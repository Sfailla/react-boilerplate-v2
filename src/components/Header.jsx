import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = props => (
	<header>
		<div>
			<h1>Expensify</h1>
			<NavLink exact to='/' activeClassName='is-active'>
				Dashboard
			</NavLink>
			<NavLink to='/create' activeClassName='is-active'>
				Create Expense
			</NavLink>
			<NavLink to='/help' activeClassName='is-active'>
				Help Page
			</NavLink>
		</div>
	</header>
);

export default Header;
