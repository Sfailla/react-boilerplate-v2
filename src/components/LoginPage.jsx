import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => {
	return (
		<Fragment>
			<div className="box-layout">
				<div className="box-layout__box">
					<h1 className="box-layout__title">Expensify</h1>
					<p>It's time to get your expenses under control</p>
					<button className="button button__login" onClick={startLogin}>
						Login with Google
					</button>
				</div>
			</div>
		</Fragment>
	);
};

const mapDispatchToProps = dispatch => {
	return {
		startLogin: () => dispatch(startLogin())
	};
};

export default connect(undefined, mapDispatchToProps)(LoginPage);
