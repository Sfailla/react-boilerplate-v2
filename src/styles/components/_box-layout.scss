.box-layout {
	background: url('/images/expensify-bg.png') no-repeat center center / cover;
	width: 100vw;
	height: calc(100vh + 5rem);
	padding: 0;
	margin: 0 0 2rem 0;
	@include flex-center(center, center);

	&__box {
		background: fade-out($color: #ffffff, $amount: .15);
		border-radius: 3px;
		text-align: center;
		width: 30rem;
		height: 30rem;
		padding: 3rem 4rem;
		@include flex-center(space-evenly, center);
		flex-direction: column;
	}

	&__title {
		padding: 0;
		margin: 0;
	}
}
