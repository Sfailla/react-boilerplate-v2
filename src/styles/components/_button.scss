.button {
	border: none;

	&__login {
		width: 80%;
		height: 4rem;
		background: #4285f4;
		color: white;
		box-shadow: 0 7px 10px rgba(black, .6);
	}

	&__logout {
		background: transparent;
		color: white;
		font-size: 2rem;
	}
}
