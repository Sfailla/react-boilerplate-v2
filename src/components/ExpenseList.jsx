import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = props => {
	const { expenses } = props;
	return (
		<Fragment>
			<h1>Expense List</h1>
			{expenses.length > 0 ? (
				expenses.map(expense => {
					return <ExpenseListItem key={expense.id} {...expense} />;
				})
			) : (
				<p>Sorry No Expenses</p>
			)}
		</Fragment>
	);
};

const mapStateToProps = state => {
	return {
		expenses: selectExpenses(state.expenses, state.filters)
	};
};

export default connect(mapStateToProps)(ExpenseList);
