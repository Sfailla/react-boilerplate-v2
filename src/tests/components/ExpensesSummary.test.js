import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render expenses summary component correctly with single expense', () => {
	const wrapper = shallow(
		<ExpensesSummary expenseCount={1} expenseTotal={195} />
	);
	expect(wrapper).toMatchSnapshot();
});

test('should render expenses summary component correctly with multiple expenses', () => {
	const wrapper = shallow(
		<ExpensesSummary expenseCount={12} expenseTotal={1500195} />
	);
	expect(wrapper).toMatchSnapshot();
});
