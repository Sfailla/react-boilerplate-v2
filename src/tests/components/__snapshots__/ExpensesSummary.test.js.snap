// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses summary component correctly with multiple expenses 1`] = `
<Fragment>
  <h1>
    Viewing 
    12
     
    expenses
     totalling 
    $1,500,195.00
  </h1>
</Fragment>
`;

exports[`should render expenses summary component correctly with single expense 1`] = `
<Fragment>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $195.00
  </h1>
</Fragment>
`;
